### DISCUSSION TITLE
(create-update-delete)Many propagation

### DISCUSSION THREAD

**Discussion by salahmedamin** [prisma/prisma#20940] [Q&A] [answered]
### Question

Hello, I am working on a project where I need to sync data between databases, and thank to Prisma's middleware feature this task became easier.

I just wanted to know if i'll have to include createMany, updateMany and deleteMany in actions array
and whether or not, prisma's (create-update-delete)Many gets mapped to an array of [ create, create, create, create, create ] and so on for the rest, when we're using *Many's actions

### Information about Prisma Schema, Client Queries and Environment (optional)

![image](https://github.com/prisma/prisma/assets/57962133/8e83d0ca-f534-48f5-bc8c-43e474505814)

**Comment by ludralph** [comment]
Hi @salahmedamin 👋 

Can you please clarify what action's array you are referring to? Also, we recommend using the [Prisma Client extensions query component type](https://www.prisma.io/docs/concepts/components/prisma-client/client-extensions/query) as an alternative to middleware. Middleware was deprecated in version 4.16.0.

**Comment by salahmedamin** [comment, marked as answer]
Hi @ludralph 
I know they're deprecated, I'll have to upgrade manually in a later version 😅
But what I mean, if I do createMany and pass required data, what does prisma do under the hood?
does it map the passed array to many prisma.create (in more optimized way) ?

**Reply by ludralph** [reply]
Hi @salahmedamin 

On calling `createMany`, Prisma uses transaction with `BEGIN` and `COMMIT` and instead of creatiing multiple `INSERT` statements we are inserting all records in a single INSERT statement.
```sql
prisma:query BEGIN
prisma:query INSERT INTO "public"."User" ("name","email") VALUES ($1,$2), ($3,$4), ($5,$6), ($7,$8) ON CONFLICT DO NOTHING
prisma:query COMMIT
```
Prisma doesn't call `create` method under the hood for `createMany` so you may need to include  `createMany`, `updateMany` and `deleteMany` in actions array.

If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏
